src/bootstrap/bootstrap.py

```diff
@@ -975,11 +975,15 @@ def build_bootstrap_cmd(self, env):
         if "RUSTFLAGS_BOOTSTRAP" in env:
             env["RUSTFLAGS"] += " " + env["RUSTFLAGS_BOOTSTRAP"]
 
-        env["PATH"] = os.path.join(self.bin_root(), "bin") + \
+        cargo_bin_path = os.path.join(self.bin_root(), "bin") + \
             os.pathsep + env["PATH"]
         if not os.path.isfile(self.cargo()):
-            raise Exception("no cargo executable found at `{}`".format(
+            if cargo_bin_path is None:
+                raise Exception("no cargo executable found at `{}`".format(
                 self.cargo()))
+            else:
+                cargo_bin_path = os.getenv("RUST_TARGET_PATH") + "rust-snapshot/bin/cargo"
```

Comment:
We don't include custom paths in the source code directly. What about adding this to "PATH" in your environment?

Reply:
I thought of including the path set by `RUST_TARGET_PATH` + "rust-snapshot/bin/cargo" and adding it as cargo bin path.
>  Why not adding it to "PATH" in your environment?

While doing rustc build I don't know the path of cargo bin beforehand.
Should I do any changes in the above patch?

Reply:
Thanks for looking into the PR.

I thought of including the path set by RUST_TARGET_PATH + "rust-snapshot/bin/cargo" and adding it as cargo bin path.

> Why not adding it to "PATH" in your environment?

While doing rustc build I don't know the path of cargo bin beforehand and the build gives error as mentioned in https://github.com/rust-lang/rust/pull/122069#issue-2170809847

Should I do any changes in the above patch?

Reply:
If you know that Cargo will be at `$RUST_TARGET_PATH + "rust-snapshot/bin/cargo"`, can't you just add that to your PATH in your build process? Or set `cargo` in `config.toml` to that

Reply:
Actually, I am running the build process with a wrapper script. `$RUST_TARGET_PATH` is set after the rust sources are generated and while build is running. If I add it in the PATH before build, it would fail as `$RUST_TARGET_PATH` is undefined path. Still, I will give it a try and check as you mentioned and get back.

Reply:
I am quite sure there are plenty of ways to solve this problem without adding hard-coded custom paths in the source code. If not, then your build flow is obviously doing something very wrong. We can't accept this change to upstream.

Reply:
Does the following changes look good to you?

Reverting the changes on line no 978. Then using that `cargo_bin_path` value to handle cases where cargo bin is not found(as in my case, since this is been handled at very last else case).

```
diff --git a/src/bootstrap/bootstrap.py b/src/bootstrap/bootstrap.py
index 0d604c0d3e5..e7224ab45b8 100644
--- a/src/bootstrap/bootstrap.py
+++ b/src/bootstrap/bootstrap.py
@@ -977,9 +977,15 @@ class RustBuild(object):
 
         env["PATH"] = os.path.join(self.bin_root(), "bin") + \
             os.pathsep + env["PATH"]
+        cargo_bin_path = os.path.join(self.bin_root(), "bin") + \
+            os.pathsep + env["PATH"]
         if not os.path.isfile(self.cargo()):
-            raise Exception("no cargo executable found at `{}`".format(
+            if cargo_bin_path is None:
+                raise Exception("no cargo executable found at `{}`".format(
                 self.cargo()))
+            else:
+                cargo_bin_path = os.getenv("RUST_TARGET_PATH") + "rust-snapshot/bin/cargo"
+                env["PATH"] = os.path.dirname(cargo_bin_path) + os.pathsep + env["PATH"]
         args = [self.cargo(), "build", "--manifest-path",
                 os.path.join(self.rust_root, "src/bootstrap/Cargo.toml")]
         args.extend("--verbose" for _ in range(self.verbose))
```

Reply:
As mentioned earlier, we can't accept hard-coded paths for specific projects like this `rust-snapshot` path. This really sounds like this would be better solved within your build flow